plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/explorer/OpenShiftExplorerContentProvider.java

```diff
@@ -192,38 +199,41 @@ public void handleAdd(int index, Object element) {
 				}
 				if(child instanceof IResourceUIModel && ResourceKind.ROUTE.equals(((IResourceUIModel) child).getResource().getKind())) {
 					updateChildrenFromViewer(parent);
+					elementsToRefresh.add(parent);
 				}
 			}
-			for (Object child : added) {
-				Object parent = getParent(child);
-				Trace.debug("Explorer add: parent: {0} / child: {1}", parent, child);
-				if(child instanceof Deployment) {
-					Deployment deployment = (Deployment)child;
+			for (Object addedElement : addedElements) {
+				Object parentElement = getParent(addedElement);
+				Trace.debug("Explorer add: parent: {0} / child: {1}", parentElement, addedElement);
+				if(addedElement instanceof Deployment) {
+					Deployment deployment = (Deployment)addedElement;
 					addDeploymentListeners(deployment);
+					elementsToRefresh.add(deployment.getParent());
 				}
-				if(child instanceof IResourceUIModel && ResourceKind.ROUTE.equals(((IResourceUIModel) child).getResource().getKind())) {
-					updateChildrenFromViewer(parent);
-				}
-				//HACK to fix JBIDE-21458
-				if(newList.size() > 0 && (parent instanceof IProjectAdapter || parent instanceof Deployment)) {
-					refreshViewer(parent);
-				}else {
-					addChildrenToViewer(parent, child);
-				}
+//				if(child instanceof IResourceUIModel && ResourceKind.ROUTE.equals(((IResourceUIModel) child).getResource().getKind())) {
```

Comment:
Is this code intended to still be commented?  This particular line explicitly checks routes because we haven chosen to display in the tree 'service - route'.  On first glance, i'm not certain the UI will reflect the case where a service exists and then a route is added.